source/Timer.cpp

```diff
@@ -0,0 +1,326 @@
+				rand = static_cast<uint32_t>(child.Value(2));
+		}
+		else if(child.Token(0) == "idle")
+		{
+			requireIdle = true;
+			// We square the max speed value here, so it can be conveniently
+			// compared to the flagship's squared velocity length below.
+			if(child.Size() > 1)
+				idleMaxSpeed = child.Value(1) * child.Value(1);
+		}
+		else if(child.Token(0) == "optional")
+			optional = true;
+		else if(child.Token(0) == "system" && child.Size() > 1)
+			system = GameData::Systems().Get(child.Token(1));
+		else if(child.Token(0) == "system")
+			systems.Load(child);
+		else if(child.Token(0) == "proximity" && child.Size() > 1)
+		{
+			proximityCenter = GameData::Planets().Get(child.Token(1));
+		}
+		else if(child.Token(0) == "proximity")
+		{
+			proximityCenters.Load(child);
+		}
+		else if(child.Token(0) == "proximity settings" && child.Size() > 1)
+		{
+			proximity = child.Value(1);
+			if(child.Size() > 2)
+			{
+				if(child.Token(2) == "close")
+					closeTo = true;
+				else if(child.Token(2) == "far")
+					closeTo = false;
+			}
+		}
+		else if(child.Token(0) == "uncloaked")
+			requireUncloaked = true;
+		else if(child.Token(0) == "reset" && child.Size() > 1)
+		{
+			const string &value = child.Token(1);
+			if(value == "leave system")
+				resetCondition = ResetCondition::LEAVE_SYSTEM;
+			else if(value == "leave zone")
+				resetCondition = ResetCondition::LEAVE_ZONE;
+			else if(value == "none")
+				resetCondition = ResetCondition::NONE;
+			else if(value == "pause")
+				resetCondition = ResetCondition::PAUSE;
+			else
+				child.PrintTrace("Skipping unrecognized attribute:");
+		}
+		else if(child.Token(0) == "repeat reset")
+			repeatReset = true;
+		else if(child.Token(0) == "reset fired")
+			resetFired = true;
+		// We keep "on timeup" as separate tokens so that it's compatible with MissionAction syntax.
+		else if(child.Token(0) == "on" && child.Size() > 1 && child.Token(1) == "timeup")
+			actions[TIMEUP].Load(child);
+		else if(child.Token(0) == "on" && child.Size() > 1 && child.Token(1) == "reset")
+			actions[RESET].Load(child);
+		else
+			child.PrintTrace("Skipping unrecognized attribute:");
+
+	}
+}
+
+
+
+// Note: the Save() function can assume this is an instantiated Timer, not a template,
+// so the time to wait will be saved fully calculated, and with any elapsed time subtracted.
+void Timer::Save(DataWriter &out) const
+{
+	// If this Timer should no longer appear in-game, don't serialize it.
+	if(isComplete)
+		return;
+
+	int64_t timeRemaining = timeToWait - timeElapsed;
+	if(!name.empty())
+		out.Write("timer", name);
+	else
+		out.Write("timer");
+	out.BeginChild();
+	{
+		out.Write("time", timeRemaining);
+		if(system)
+			out.Write("system", system->Name());
+		else if(!systems.IsEmpty())
+		{
+			out.Write("system");
+			systems.Save(out);
+		}
+		if(requireIdle)
+		{
+			if(idleMaxSpeed == DEFAULT_MAX_SPEED)
+				out.Write("idle");
+			else
+				out.Write("idle", sqrt(idleMaxSpeed));
```

Comment:
No harm in just outputting the default.
```suggestion
			out.Write("idle", sqrt(idleMaxSpeed));
```